{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import pandas as pd\n",
    "import json\n",
    "import itertools\n",
    "import numpy as np\n",
    "import json\n",
    "from statsmodels.stats.contingency_tables import mcnemar\n",
    "from scipy.stats import fisher_exact\n",
    "\n",
    "tad_path = 'tad/GSE63525_K562_{}_TAD.txt'\n",
    "tss_path = 'refTSS_v3.1_human_annotation.txt'\n",
    "annot_path = 'refTSS_v3.1_human_coordinate.hg38.bed'\n",
    "cont_path = \"K562_tss_survival.json\"\n",
    "result_path = 'GWAS_results.txt'\n",
    "pair_path = \"pairs_by_chrom.json\"\n",
    "\n",
    "tads = []\n",
    "\n",
    "tad_file = open('GSE63525_K562_domains.bed')\n",
    "header = ['Start', 'End', 'Cen']\n",
    "\n",
    "chrom_tad_path = 'tad/GSE63525_K562_chr{}_TAD.txt'\n",
    "\n",
    "sqtl_path = 'LAML_CompleteSurvival_sQTLs.csv'\n",
    "\n",
    "RESOLUTION = 10000\n",
    "CHROMOSOMES = [\"chr1\",\"chr2\",\"chr3\",\"chr4\",\"chr5\",\"chr6\",\"chr7\",\"chr8\",\"chr9\",\"chr10\",\"chr11\",\"chr12\",\"chr13\",\"chr14\",\"chr15\",\"chr16\",\"chr17\",\"chr18\",\"chr19\",\"chr20\",\"chr21\",\"chr22\",\"chrX\",\"chrY\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def symbol_to_tss_id(symbol):\n",
    "    return [id for (sym,id) in zip(safe_tss_symbols,safe_tss_ids) if sym == symbol]\n",
    "\n",
    "def tss_id_to_pos(id):\n",
    "    return safe_tss_pos[tss_pos_ids.index(id)]\n",
    "\n",
    "def is_between(list_sup,list_sub):\n",
    "    #Big list first small list later!\n",
    "    return min(list_sub) >= min(list_sup) and max(list_sub) <= max(list_sup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "check_chrom = '1'\n",
    "temp = []\n",
    "for line in tad_file:\n",
    "    current_line = line.split('\\t')[:3]\n",
    "    bin_values = (\n",
    "        int(int(current_line[1]) / RESOLUTION), #Start\n",
    "        int(int(current_line[2]) / RESOLUTION), #End\n",
    "        0                                       #Cen\n",
    "    )\n",
    "    if check_chrom != current_line[0]:\n",
    "        with open(tad_path.format(check_chrom), 'wt', newline ='') as file:\n",
    "            writer = csv.writer(file, delimiter='\\t')\n",
    "            writer.writerow(i for i in header)\n",
    "            for j in temp:\n",
    "                writer.writerow(j)\n",
    "        print(check_chrom,'Done')\n",
    "        check_chrom = current_line[0]\n",
    "        temp = [bin_values]\n",
    "        \n",
    "\n",
    "    else:\n",
    "        temp.append(bin_values)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_by_chrom = {}\n",
    "for chrom in CHROMOSOMES:\n",
    "    pairs_by_chrom[chrom] = []\n",
    "\n",
    "prad = pd.read_csv(sqtl_path,delimiter=';')\n",
    "prad_snp_pos = prad['SNP position'].tolist()\n",
    "prad_gene_symbols = prad['Gene'].tolist()\n",
    "\n",
    "tss_ids_df = pd.read_csv(tss_path,delimiter = '\\t')\n",
    "tss_ids_df.dropna(subset = [\"Gene_symbol\"], inplace=True)\n",
    "safe_tss_symbols = tss_ids_df['Gene_symbol'].tolist()\n",
    "safe_tss_ids = tss_ids_df['#refTSSID'].tolist()\n",
    "\n",
    "tss_pos_df = pd.read_csv(annot_path,delimiter = '\\t',header = None)\n",
    "tss_pos_chroms, safe_tss_pos, tss_pos_ids = tss_pos_df[0].tolist(),tss_pos_df[1].tolist(),tss_pos_df[3].tolist()\n",
    "\n",
    "def symbol_to_tss_id(symbol):\n",
    "    return [id for (sym,id) in zip(safe_tss_symbols,safe_tss_ids) if sym == symbol]\n",
    "\n",
    "def tss_id_to_pos(id):\n",
    "    return safe_tss_pos[tss_pos_ids.index(id)]\n",
    "    \n",
    "symbol_snp_pos = {}\n",
    "for snp_pos,gene_symbol in zip(prad_snp_pos,prad_gene_symbols):\n",
    "    if gene_symbol not in symbol_snp_pos.keys():\n",
    "        symbol_snp_pos[gene_symbol] = [int(snp_pos.split(':')[1])]\n",
    "    else:\n",
    "        symbol_snp_pos[gene_symbol].append(int(snp_pos.split(':')[1]))\n",
    "\n",
    "pairs = {}\n",
    "count = 0\n",
    "for gene_symbol_key in symbol_snp_pos:\n",
    "    tss_ids =  symbol_to_tss_id(gene_symbol_key)\n",
    "    tss_positions = [tss_id_to_pos(tss_id) for tss_id in tss_ids]\n",
    "    pairs[gene_symbol_key] = list(itertools.product(tss_positions,symbol_snp_pos[gene_symbol_key]))\n",
    "    count += 1\n",
    "    print(gene_symbol_key,'pairs created\\t',round(count/4990*100,2),'%')\n",
    "\n",
    "chrom_genes = {}\n",
    "for pos,gene in zip(prad_snp_pos,prad_gene_symbols):\n",
    "    if gene not in chrom_genes.keys():\n",
    "        chrom_genes[gene] = pos.split(':')[0]\n",
    "\n",
    "count = 0\n",
    "total_size = len(pairs)\n",
    "for gene in pairs:\n",
    "    for pair in pairs[gene]:\n",
    "        pairs_by_chrom[chrom_genes[gene]].append(pair)\n",
    "    count += 1\n",
    "    print(gene,'pairs added to the result\\t',round(count/total_size*100,2),'%')\n",
    "\n",
    "\n",
    "with open(pair_path, \"w\") as outfile: \n",
    "    json.dump(pairs_by_chrom, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(pair_path) as data_file:\n",
    "    data = json.load(data_file)\n",
    "\n",
    "#Create result dictionary empty\n",
    "result_by_chrom = {}\n",
    "for chrom in CHROMOSOMES:\n",
    "    result_by_chrom[chrom] = [[0,0],[0,0]]\n",
    "\n",
    "for key in data:\n",
    "    try:\n",
    "        tad_data = pd.read_csv(chrom_tad_path.format(key),delimiter='\\t')\n",
    "        tad_list = np.array([tad_data['Start'].tolist(),tad_data['End'].tolist()]).T.tolist()\n",
    "\n",
    "        true_pairs = [(int(pair[0]/RESOLUTION),int(pair[1]/RESOLUTION)) for pair in data[key]]  #(a,b) true (3,5)\n",
    "        pseudo_pairs = [(pair[1],pair[1]-pair[0]+pair[1]) for pair in true_pairs]   #(b,b - a + b) pseudo (5,7)\n",
    "\n",
    "        count = 0\n",
    "        total_size = len(true_pairs)\n",
    "\n",
    "        for pair in true_pairs:\n",
    "            not_found = True\n",
    "            count += 1\n",
    "            print(key,'True Pairs\\t\\t',round(count/total_size*100,2),'%')\n",
    "            for tad in tad_list:\n",
    "                if is_between(tad,pair):\n",
    "                    result_by_chrom[key][0][0] += 1 #TrueYes\n",
    "                    not_found = False\n",
    "                    break\n",
    "            if not_found:\n",
    "                result_by_chrom[key][0][1] += 1 #TrueNo\n",
    "\n",
    "        count = 0\n",
    "        for pair in pseudo_pairs:\n",
    "            not_found = True\n",
    "            count += 1\n",
    "            print(key,'Pseudo Pairs\\t\\t',round(count/total_size*100,2),'%')\n",
    "            for tad in tad_list:\n",
    "                if is_between(tad,pair):\n",
    "                    result_by_chrom[key][1][0] += 1 #PseudoYes\n",
    "                    not_found = False\n",
    "                    break\n",
    "            if not_found:\n",
    "                result_by_chrom[key][1][1] += 1 #PseudoNo\n",
    "    except FileNotFoundError:\n",
    "        print('TAD for',key,'does not exist')\n",
    "        \n",
    "            \n",
    "            \n",
    "print('Started Saving')\n",
    "with open(cont_path, \"w\") as outfile: \n",
    "    json.dump(result_by_chrom, outfile)\n",
    "print('DONE')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(cont_path) as json_file:\n",
    "    matrices = json.load(json_file)\n",
    "    \n",
    "result_file = open(result_path,'w')\n",
    "for chrom in CHROMOSOMES:\n",
    "    try:\n",
    "        cont_matrix = matrices[chrom]\n",
    "        print(chrom,'Contingency Matrix:','\\n',cont_matrix[0],'\\n',cont_matrix[1],'\\n---',file = result_file)\n",
    "        print(chrom,'Mcnemar Results:\\n---',file = result_file)\n",
    "        print(mcnemar(cont_matrix),'\\n---',file = result_file)\n",
    "        oddsr, p = fisher_exact(cont_matrix)\n",
    "        print(chrom,'Fisher Exact Results:\\n---',file = result_file)\n",
    "        print('Odds Ratio\\t',oddsr,file = result_file)\n",
    "        print('pvalue fish\\t',p,file = result_file)\n",
    "        print('\\n--------\\n',file = result_file)\n",
    "    except KeyError:\n",
    "        print(chrom,'doesnt exist\\n---')\n",
    "json_file.close()\n",
    "result_file.close()"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
